torch/_inductor/codegen/cpp.py

```diff
@@ -5318,34 +5343,49 @@ def max_parallel_depth(self):
         1) Levels without splitting and
         2) All reduction or non-reduction levels
         When the loop is split at the top level, the max depth is 1.
+        When the range of the first inner loop is much larger than the range of all outer loops,
+        change the starting depth of parallelization to the first inner loop.
         """
         if self.loops is None:
-            return 0
+            return ParallelDepth(parallel_depth=0, start_depth=0)
 
+        start_depth = 0
         max_depth = 0
         is_reduction = self.loops[0].is_reduction
+        loop_sizes = sympy.Integer(1)
         for loop in self.loops:
             if loop.is_reduction != is_reduction:
                 break
+            loop_sizes = loop_sizes * loop.size
             max_depth += 1
-        return max_depth
 
-    def is_reduction_only(self):
-        """
-        Whether all the loops are for reduction. Reduction loops
-        are always the inner most ones.
-        """
-        return self.loops is not None and self.loops[0].is_reduction
+        # When the range of the first inner loop is much larger than the range of all outer loops,
+        # change `start_depth` to the first inner loop and recalculate `max_depth`.
+        if (
+            max_depth < len(self.loops)
+            and isinstance(loop_sizes, sympy.Integer)
+            and isinstance(self.loops[max_depth].size, sympy.Integer)
+            and loop_sizes * 300 < self.loops[max_depth].size
+        ):
+            start_depth = max_depth
+            max_depth = 0
+            is_reduction = self.loops[start_depth].is_reduction
+            for i in range(start_depth, len(self.loops)):
+                if self.loops[i].is_reduction != is_reduction:
+                    break
+                max_depth += 1
+        return ParallelDepth(parallel_depth=max_depth, start_depth=start_depth)
 
     def mark_parallel(self, par_depth):
         assert (
-            par_depth <= self.max_parallel_depth()
+            par_depth <= self.max_parallel_depth().parallel_depth
         ), "Parallel depth cannot exceed the maximal allowed parallel depth"
         assert self.loops is not None
         assert len(self.loops) >= par_depth
-        loop = self.loops[0]
+        start_depth = self.max_parallel_depth().start_depth
```

Comment:
Should we use `self.max_parallel_depth().start_depth` or pass `par_depth` from outside?

Reply:
For the current implementation, there is no difference between the two, but pass `par_depth `from outside does seem more reasonable.

Reply:
Modified, thanks for your comment!